juju/juju
# [JUJU-1379] Preserve order of relations in bundles

When deploying a bundle, Juju was adding the relations in reverse order. This meant when you re-exported the bundle, the relations would be reordered, causing failures in the bash tests.

I tracked this down to the `toposortFlatten` function in `core/bundle/changes/changes.go`, which was used to sort bundle changes to be applied. This function was added in #13259. The sorting algorithm used was not *stable*, i.e. if you have two changes for which the order doesn't matter, their original order may not be preserved. This was causing the above issue.

I rewrote this function using a simplified, *stable* topological sorting algorithm, and added a unit test in `core/bundle/changes/changessort_test.go` to test stability for relations.

Fingers crossed this should fix `test-deploy-test-deploy-bundles-aws`.

## QA steps

Save this file as `testbdl.yml`:
```yml
series: focal
applications:
  src:
    charm: ./acceptancetests/repository/charms/dummy-source
  snk1:
    charm: ./acceptancetests/repository/charms/dummy-sink
  snk2:
    charm: ./acceptancetests/repository/charms/dummy-sink
  snk3:
    charm: ./acceptancetests/repository/charms/dummy-sink
relations:
- - src:sink
  - snk1:source
- - src:sink
  - snk2:source
- - src:sink
  - snk3:source
```

Run the following commands in the root directory of the Juju source tree (assuming you already have a controller bootstrapped and selected):
```sh
juju add-model m
juju deploy ./testbdl.yml
juju export-bundle --filename testbdl2.yml
```

Open `testbdl2.yml` and verify that the relations are in the same order.
